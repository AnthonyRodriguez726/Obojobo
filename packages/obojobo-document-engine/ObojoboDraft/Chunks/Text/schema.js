import { Block } from 'slate'
import { CHILD_TYPE_INVALID } from 'slate-schema-violations'

const TEXT_LINE_NODE = 'ObojoboDraft.Chunks.Text.TextLine'

const schema = {
	blocks: {
		'ObojoboDraft.Chunks.Text': {
			nodes: [
				{
					match: [{ type: TEXT_LINE_NODE }],
					min: 1
				}
			],
			normalize: (editor, error) => {
				const { node, child, index } = error
				switch (error.code) {
					case CHILD_TYPE_INVALID: {
						// Allow inserting of new nodes by unwrapping unexpected blocks at end and beginning
						const isAtEdge = index === node.nodes.size - 1 || index === 0
						if (child.object === 'block' && isAtEdge) {
							return editor.unwrapNodeByKey(child.key)
						}

						return editor.wrapBlockByKey(child.key, {
							type: TEXT_LINE_NODE,
							data: { indent: 0 }
						})
					}
					case 'child_min_invalid': {
						const block = Block.create({
							type: TEXT_LINE_NODE,
							data: { indent: 0 }
						})
						return editor.insertNodeByKey(node.key, index, block)
					}
				}
			}
		},
		'ObojoboDraft.Chunks.Text.TextLine': {
			nodes: [{ match: [{ object: 'text' }] }],
			normalize: (editor, error) => {
				const { node, child, index } = error
				switch (error.code) {
					case CHILD_TYPE_INVALID: {
						// Allow inserting of new nodes by unwrapping unexpected blocks at end and beginning
						const isAtEdge = index === node.nodes.size - 1 || index === 0
						if (child.object === 'block' && isAtEdge) {
							return editor.unwrapNodeByKey(child.key)
						}
					}
				}
			}
		}
	}
}

export default schema
