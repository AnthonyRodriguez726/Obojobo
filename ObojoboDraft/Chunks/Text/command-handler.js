let CommandHandler
let { Editor } = window
import Common from 'Common'

let { TextGroupCommandHandler } = Editor.chunk.textChunk
let { TextGroupSelection } = Common.textGroup

export default (CommandHandler = class CommandHandler extends TextGroupCommandHandler {
	onEnter(selection, chunk, shiftKey) {
		chunk.markDirty()

		let tgs = new TextGroupSelection(chunk, selection.virtual)
		let { textGroup } = chunk.modelState

		if (tgs.start.text.length !== 0) {
			chunk.splitText()
			return
		}

		// If splitting an empty one item paragraph create a new paragraph below
		if (textGroup.length === 1) {
			let sibChunk = chunk.clone()
			sibChunk.modelState.textGroup.init(1)

			chunk.addChildAfter(sibChunk)

			// TextGroupSelection.setCaretToGroupStart sibChunk, selection.virtual
			return sibChunk.selectStart()

			// If splitting an empty textGroup item that's not a one item paragraph
			// then split into three paragraphs:
		} else {
			let splitChildren
			return (splitChildren = chunk.split())
		}
	}
})
