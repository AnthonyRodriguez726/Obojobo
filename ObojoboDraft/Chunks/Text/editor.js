import React from 'react'
import { Block } from 'slate'
import { CHILD_REQUIRED, CHILD_TYPE_INVALID } from 'slate-schema-violations'

import TextUtil from '../../../src/scripts/oboeditor/util/text-util'

const TEXT_NODE = 'ObojoboDraft.Chunks.Text'
const TEXT_LINE_NODE = 'ObojoboDraft.Chunks.Text.TextLine'

const Line = props => {
	return (
		<span
			className={'text align-' + props.node.data.get('align')}
			{...props.attributes}
			data-indent={props.node.data.get('indent')}
		>
			{props.children}
		</span>
	)
}

const Node = props => {
	return (
		<div className={'component'}>
			<div className={'text-chunk obojobo-draft--chunks--single-text pad'}>{props.children}</div>
		</div>
	)
}

const isType = change => {
	return change.value.blocks.some(block => {
		return !!change.value.document.getClosest(block.key, parent => {
			return parent.type === TEXT_NODE
		})
	})
}

const insertNode = change => {
	change
		.insertBlock(TEXT_NODE)
		.collapseToStartOfNextText()
		.focus()
}

const slateToObo = node => {
	const json = {}
	json.id = node.key
	json.type = node.type
	json.content = node.data.get('content') || {}

	json.content.textGroup = []
	node.nodes.forEach(line => {
		const textLine = {
			text: { value: line.text, styleList: [] },
			data: { indent: line.data.get('indent') }
		}

		line.nodes.forEach(text => {
			TextUtil.slateToOboText(text, textLine)
		})

		json.content.textGroup.push(textLine)
	})
	json.children = []

	return json
}

const oboToSlate = node => {
	const json = {}
	json.object = 'block'
	json.key = node.id
	json.type = node.type
	json.data = { content: {} }

	json.nodes = []
	node.content.textGroup.forEach(line => {
		const indent = line.data ? line.data.indent : 0
		const align = line.data ? line.data.align : 0
		const codeLine = {
			object: 'block',
			type: TEXT_LINE_NODE,
			data: { indent, align },
			nodes: [
				{
					object: 'text',
					leaves: TextUtil.parseMarkings(line)
				}
			]
		}

		json.nodes.push(codeLine)
	})

	return json
}

const plugins = {
	onKeyDown(event, change) {
		const isText = isType(change)

		// Enter
		if (isText && event.key === 'Enter') {
			event.preventDefault()
			const last = change.value.endBlock
			// Double Enter
			if (last.text === '') {
				change.removeNodeByKey(last.key)
				change.splitBlock(2)
				return true
			}
			change.insertBlock({ type: TEXT_LINE_NODE, data: { indent: 0 } })
			return true
		}

		// Shift Tab
		if (isText && event.key === 'Tab' && event.shiftKey) {
			event.preventDefault()
			change.value.blocks.forEach(block => {
				let newIndent = block.data.get('indent') - 1
				if (newIndent < 1) newIndent = 0

				return change.setNodeByKey(block.key, {
					data: { indent: newIndent }
				})
			})
			return true
		}

		// Tab indent
		if (isText && event.key === 'Tab') {
			event.preventDefault()
			change.value.blocks.forEach(block =>
				change.setNodeByKey(block.key, {
					data: { indent: block.data.get('indent') + 1 }
				})
			)
			return true
		}
	},
	renderNode(props) {
		switch (props.node.type) {
			case TEXT_NODE:
				return <Node {...props} />
			case TEXT_LINE_NODE:
				return <Line {...props} />
		}
	},
	schema: {
		blocks: {
			'ObojoboDraft.Chunks.Text': {
				nodes: [{ types: [TEXT_LINE_NODE] }],
				normalize: (change, violation, { node, child, index }) => {
					switch (violation) {
						case CHILD_TYPE_INVALID: {
							return change.wrapBlockByKey(child.key, {
								type: TEXT_LINE_NODE,
								data: { indent: 0 }
							})
						}
						case CHILD_REQUIRED: {
							const block = Block.create({
								type: TEXT_LINE_NODE,
								data: { indent: 0 }
							})
							return change.insertNodeByKey(node.key, index, block)
						}
					}
				}
			},
			'ObojoboDraft.Chunks.Text.TextLine': {
				nodes: [{ objects: ['text'] }]
			}
		}
	}
}

const Text = {
	components: {
		Node,
		Line
	},
	helpers: {
		insertNode,
		slateToObo,
		oboToSlate
	},
	plugins
}

export default Text
