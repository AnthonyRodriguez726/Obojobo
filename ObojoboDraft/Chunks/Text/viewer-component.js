import './viewer-component.scss'

import React from 'react'

import Common from 'Common'

const { OboComponent } = Common.components
const { TextGroupEl } = Common.chunk.textChunk
const { TextChunk } = Common.chunk

const Text = props => (
	<OboComponent model={props.model} moduleData={props.moduleData}>
		<TextChunk className="obojobo-draft--chunks--single-text pad">
			{props.model.modelState.textGroup.items.map((textItem, index) => (
				<TextGroupEl textItem={textItem} groupIndex={index} key={index} parentModel={props.model} />
			))}
		</TextChunk>
	</OboComponent>
)

export default Text
