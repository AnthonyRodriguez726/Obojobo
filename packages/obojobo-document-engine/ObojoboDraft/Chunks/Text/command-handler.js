const { Editor } = window
import Common from 'Common'

const { TextGroupCommandHandler } = Editor.chunk.textChunk
const { TextGroupSelection } = Common.textGroup

export default class CommandHandler extends TextGroupCommandHandler {
	onEnter(selection, chunk) {
		chunk.markDirty()

		const tgs = new TextGroupSelection(chunk, selection.virtual)
		const { textGroup } = chunk.modelState

		if (tgs.start.text.length !== 0) {
			chunk.splitText()
			return
		}

		// If splitting an empty one item paragraph create a new paragraph below
		if (textGroup.length === 1) {
			const sibChunk = chunk.clone()
			sibChunk.modelState.textGroup.init(1)

			chunk.addChildAfter(sibChunk)

			// TextGroupSelection.setCaretToGroupStart sibChunk, selection.virtual
			return sibChunk.selectStart()

			// If splitting an empty textGroup item that's not a one item paragraph
			// then split into three paragraphs:
		} else {
			return chunk.split()
		}
	}
}
