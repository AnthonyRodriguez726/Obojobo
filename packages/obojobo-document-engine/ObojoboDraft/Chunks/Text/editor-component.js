import React from 'react'

const Text = props => (
	<div className={'text-chunk obojobo-draft--chunks--single-text pad'}>{props.children}</div>
)

export default Text
