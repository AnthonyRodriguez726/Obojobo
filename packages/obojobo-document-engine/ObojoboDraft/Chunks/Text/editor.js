import React from 'react'
import { Block } from 'slate'
import { CHILD_REQUIRED, CHILD_TYPE_INVALID } from 'slate-schema-violations'

import TextUtil from '../../../src/scripts/oboeditor/util/text-util'

import emptyNode from './empty-node.json'
import Icon from './icon'

import OboEditorStore from '../../../src/scripts/oboeditor/store'

const TEXT_NODE = 'ObojoboDraft.Chunks.Text'
const TEXT_LINE_NODE = 'ObojoboDraft.Chunks.Text.TextLine'

const Line = props => {
	return (
		<span
			className={'text align-' + props.node.data.get('align')}
			data-indent={props.node.data.get('indent')}
		>
			{props.children}
		</span>
	)
}

const Node = props => {
	return (
		<div className={'text-chunk obojobo-draft--chunks--single-text pad'}>
			{props.children}
		</div>
	)
}

const isType = change => {
	return change.value.blocks.some(block => {
		return !!change.value.document.getClosest(block.key, parent => {
			return parent.type === TEXT_NODE
		})
	})
}

const insertNode = change => {
	change
		.insertBlock(emptyNode)
		.focus()
}

const slateToObo = node => {
	const json = {}
	json.id = node.key
	json.type = node.type
	json.content = {}

	json.content.textGroup = []
	node.nodes.forEach(line => {
		const textLine = {
			text: { value: line.text, styleList: [] },
			data: { indent: line.data.get('indent'), align: line.data.get('align') }
		}

		line.nodes.forEach(text => {
			TextUtil.slateToOboText(text, textLine)
		})

		json.content.textGroup.push(textLine)
	})
	json.children = []

	return json
}
const oboToSlate = node => {
	const json = {}
	json.object = 'block'
	json.key = node.id
	json.type = node.type
	json.data = { content: {} }

	json.nodes = []
	node.content.textGroup.forEach(line => {
		const indent = line.data ? line.data.indent : 0
		const align = line.data ? line.data.align : 0
		const textLine = {
			object: 'block',
			type: TEXT_LINE_NODE,
			data: { indent, align },
			nodes: [
				{
					object: 'text',
					leaves: TextUtil.parseMarkings(line)
				}
			]
		}

		json.nodes.push(textLine)
	})

	return json
}

const plugins = {
	onKeyDown(event, change) {
		const isText = isType(change)
		if (!isText) return

		// Delete empty text node
		if (event.key === 'Backspace' || event.key === 'Delete') {
			const last = change.value.endBlock
			let parent
			change.value.blocks.forEach(block => {
				parent = change.value.document.getClosest(block.key, parent => {
					return parent.type === TEXT_NODE
				})
			})

			if (last.text === '' && parent.nodes.size === 1) {
				event.preventDefault()
				change.removeNodeByKey(parent.key)
				return true
			}
		}

		// Enter
		if (event.key === 'Enter') {
			event.preventDefault()
			const last = change.value.endBlock
			// Double Enter
			if (last.text === '') {
				change.removeNodeByKey(last.key)
				change.splitBlock(2)
				return true
			}

			return
		}

		// Shift Tab
		if (event.key === 'Tab' && event.shiftKey) {
			event.preventDefault()
			change.value.blocks.forEach(block => {
				let newIndent = block.data.get('indent') - 1
				if (newIndent < 1) newIndent = 0

				return change.setNodeByKey(block.key, {
					data: { indent: newIndent }
				})
			})
			return true
		}

		// Alt Tab
		if (event.key === 'Tab' && event.altKey) {
			event.preventDefault()
			change.value.blocks.forEach(block =>
				change.setNodeByKey(block.key, {
					data: { indent: block.data.get('indent') + 1 }
				})
			)
			return true
		}

		// Tab insert
		if (event.key === 'Tab') {
			event.preventDefault()
			change.insertText('\t')
			return true
		}
	},
	renderNode(props) {
		switch (props.node.type) {
			case TEXT_NODE:
				return <Node {...props} {...props.attributes} />
			case TEXT_LINE_NODE:
				return <Line {...props} {...props.attributes} />
		}
	},
	renderPlaceholder(props) {
		const { node } = props
		if (node.object !== 'block' || node.type !== TEXT_LINE_NODE) return
		if (node.text !== '') return

		return (
			<span className={'placeholder align-' + node.data.get('align')} contentEditable={false}>
				{'Type Your Text Here'}
			</span>
		)
	},
	schema: {
		blocks: {
			'ObojoboDraft.Chunks.Text': {
				nodes: [
					{
						match: [{ type: TEXT_LINE_NODE }],
						min: 1
					}
				],
				normalize: (change, error) => {
					const { node, child, index } = error
					switch (error.code) {
						case CHILD_TYPE_INVALID: {
							// Allow inserting of new nodes by unwrapping unexpected blocks at end and beginning
							const isAtEdge = index === node.nodes.size - 1 || index === 0
							if (child.object === 'block' && isAtEdge) {
								return change.unwrapNodeByKey(child.key)
							}

							return change.wrapBlockByKey(child.key, {
								type: TEXT_LINE_NODE,
								data: { indent: 0 }
							})
						}
						case CHILD_REQUIRED: {
							const block = Block.create({
								type: TEXT_LINE_NODE,
								data: { indent: 0 }
							})
							return change.insertNodeByKey(node.key, index, block)
						}
					}
				}
			},
			'ObojoboDraft.Chunks.Text.TextLine': {
				nodes: [{ match: [{ object: 'text' }] }],
				normalize: (change, error) => {
					const { node, child, index } = error
					switch (error.code) {
						case CHILD_TYPE_INVALID: {
							// Allow inserting of new nodes by unwrapping unexpected blocks at end and beginning
							const isAtEdge = index === node.nodes.size - 1 || index === 0
							if (child.object === 'block' && isAtEdge) {
								return change.unwrapNodeByKey(child.key)
							}
						}
					}
				}
			}
		}
	}
}

const Text = {
	name: TEXT_NODE,
	components: {
		Node,
		Line,
		Icon
	},
	helpers: {
		insertNode,
		slateToObo,
		oboToSlate
	},
	json: {
		emptyNode
	},
	plugins
}

OboEditorStore.registerModel('ObojoboDraft.Chunks.Text', {
	name: 'Text',
	icon: Icon,
	isInsertable: true,
	componentClass: Node,
	insertJSON: emptyNode,
	plugins
})

export default Text
