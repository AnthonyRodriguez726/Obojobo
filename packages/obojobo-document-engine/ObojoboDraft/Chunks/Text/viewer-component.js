import './viewer-component.scss'

import Common from 'Common'

let { OboComponent } = Common.components
let { TextGroupEl } = Common.chunk.textChunk
let { TextChunk } = Common.chunk
let { Dispatcher } = Common.flux

export default class Text extends React.Component {
	render() {
		let texts = this.props.model.modelState.textGroup.items.map(
			((textItem, index) => {
				return (
					<TextGroupEl
						textItem={textItem}
						groupIndex={index}
						key={index}
						parentModel={this.props.model}
					/>
				)
			}).bind(this)
		)

		return (
			<OboComponent model={this.props.model} moduleData={this.props.moduleData}>
				<TextChunk className="obojobo-draft--chunks--single-text pad">
					{texts}
				</TextChunk>
			</OboComponent>
		)
	}
}
