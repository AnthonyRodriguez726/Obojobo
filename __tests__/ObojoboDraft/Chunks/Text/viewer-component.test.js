import React from 'react'
import renderer from 'react-test-renderer'

import Text from '../../../../ObojoboDraft/Chunks/Text/viewer-component'
import OboModel from '../../../../__mocks__/_obo-model-with-chunks'

describe('Text', () => {
	test('Text component', () => {
		let model = OboModel.create({
			id: 'id',
			type: 'ObojoboDraft.Chunks.Text',
			content: {
				textGroup: [
					{
						text: {
							value: 'Example Text'
						}
					}
				]
			}
		})
		let moduleData = {
			focusState: {}
		}

		const component = renderer.create(<Text model={model} moduleData={moduleData} />)
		let tree = component.toJSON()

		expect(tree).toMatchSnapshot()
	})
})
