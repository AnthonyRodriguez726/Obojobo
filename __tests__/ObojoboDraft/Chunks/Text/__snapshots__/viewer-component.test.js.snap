// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Text component 1`] = `
<div
  className="component"
  data-focussed={false}
  data-id="id"
  data-obo-component={true}
  data-type="ObojoboDraft.Chunks.Text"
  id="obo-id"
  model={
    Object {
      "children": null,
      "content": Object {
        "textGroup": Array [
          Object {
            "data": Object {
              "align": "left",
              "indent": 0,
            },
            "text": Object {
              "styleList": null,
              "value": "Example Text",
            },
          },
        ],
      },
      "id": "id",
      "index": 0,
      "metadata": Object {},
      "type": "ObojoboDraft.Chunks.Text",
    }
  }
  moduleData={
    Object {
      "focusState": Object {},
    }
  }
  tag="div"
>
  <div
    className="text-chunk obojobo-draft--chunks--single-text pad"
  >
    <span
      className="text align-left"
      data-group-index={0}
      data-indent={0}
    >
      <span>
        <span>
          <span>
            Example Text
          </span>
        </span>
      </span>
    </span>
  </div>
</div>
`;
